README.md
<h1 align="center">Cusys based on react</h1>

[![antd v4.22.7](https://img.shields.io/badge/antd-v4.22.7-f1b8e4.svg)](https://ant.design/docs/spec/introduce) 
[![react v18.2.0](https://img.shields.io/badge/react-v18.2.0-f2debd.svg)](https://react.dev/learn) 
[![typescript v4.7.4](https://img.shields.io/badge/typescript-v4.7.4-f1f1b8.svg)](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html) 

## What is it?
**AdvancedReact** 包含了在线转换工具、权限管理页面、角色管理页面和技术分享页面。
## Table of Contents
- [Technologies](#Technologies)
- [Development](#Development)
- [Production](#Production)
- [Modules](#)
  - online conversion tools
  - permission management
  - pages and role management pages
- [links](#)
  - [开发计划](docs/TodoList.md)
  - [历史记录](docs/CHANGELOG.md)

## Technologies
- [React](https://react.dev/learn)
- [TypeScript](https://www.typescriptlang.org/)
- [Antd Design](https://ant.design/docs/react/introduce-cn)
## Development
- pull repository
```sh
git init
git remote add origin git@github.com:imc-ux/AdvancedReact.git
git pull origin main
```
- dependencies installation
```sh
npm install
```
- local testing
```sh
npm run server
```
- push your changes
```sh
git init
git add file
git commit -m "modify information"
git branch -M main
git remote add origin git@github.com:imc-ux/AdvancedReact.git
git push -u origin main
```
本地访问路径：<br>
在线转换工具：[http://127.0.0.1:8080/cusys/react/index.html?#/main](http://127.0.0.1:8080/cusys/react/index.html?#/main)<br>
权限管理：[http://127.0.0.1:8080/cusys/react/index.html?#/permission](http://127.0.0.1:8080/cusys/react/index.html?#/permission)<br>
角色管理：[http://127.0.0.1:8080/cusys/react/index.html?#/role](http://127.0.0.1:8080/cusys/react/index.html?#/role)<br>
技术分享：[http://127.0.0.1:8080/cusys/react/index.html?#/share](http://127.0.0.1:8080/cusys/react/index.html?#/share)
## Production
```sh
npm run build
```
----
[Go to Top](#Table-of-Contents)


docs/CHANGELOG.md
## [1.0.1](https://github.com/imc-ux/AdvancedReact/releases/tag/v1.0.1) (2023-10-30)
### Features
-#10
-#7

## [1.0.0](https://github.com/imc-ux/AdvancedReact/releases/tag/v1.0.0) (2023-09-21)
### Features
- 在线转换工具
  - Json格式化
  - Base64
  - 时间戳与北京时间转换
  - MD5加密
  - SHA加密
  - 编码
- 权限管理
- 角色管理
- 技术分享

### Bug Fixes
- 无


src/util/constants.ts
export enum RoleManager {
  GetUserRoleList = 'getUserRoleList',
  GetRoleList = 'getRoleList',
  CreateRoleToUser = 'createRoleToUser',
  CreateRoleInfo = 'createRoleInfo',
  DeleteRoleFromUser = 'deleteRoleFromUser',
  UpdateRoleInfo = 'updateRoleInfo',
  GetUserActivePermission = 'getUserActivePermission',
}

export enum PermissionManager {
  GetRolePermissionList = 'getRolePermissionList',
  GetPermissionList = 'getPermissionList',
  CreatePermissionInfo = 'createPermissionInfo',
  UpdatePermissionInfo = 'updatePermissionInfo',
  CreatePermissionToRole = 'createPermissionToRole',
  DeletePermissionFromRole = 'deletePermissionFromRole',
  GetUserActivePermission = 'getUserActivePermission',
}

export enum ShareManager {
  GetUserList = 'getUserList',
  GetCategoryList = 'getCategoryList',
  GetTitleList = 'getTitleList',
  ClickShareTimes = 'clickShareTimes',
  DeleteMyShare = 'deleteMyShare',
  ToTopShare = 'toTopShare',
  AddMyShare = 'addMyShare',
  MyShareDetail = 'myShareDetail',
  ModifyMyShare = 'modifyMyShare',
  GetUserActivePermission = 'getUserActivePermission',
}


src/util/requestUtil.ts
const serviceUrl = 'http://109.14.20.45:6636/ux';
const testUrl = '/lxz';
const url = document.location.protocol + '//' + document.location.host;
const isLocal = url.includes('127.0.0.1') || url.includes('localhost');

export function callService(method: string, parameters?: any, requestType: string = 'get', isDecode = true) {
  const xhr: XMLHttpRequest = new XMLHttpRequest();
  let requestURL: string = `${isLocal ? testUrl : serviceUrl}/${method}`;
  if (requestType === 'get' && parameters) {
    if (isDecode) {
      requestURL += `?jsonString=${encodeURIComponent(encodeURIComponent(JSON.stringify(parameters)))}`;
    } else {
      requestURL += `?jsonString=${encodeURIComponent(encodeURIComponent(parameters))}`;
    }
  }
  xhr.open(requestType, requestURL, true);
  xhr.setRequestHeader('Content-type', 'application/json');

  if (requestType === 'post') {
    if (typeof parameters === 'object') {
      xhr.send(JSON.stringify(parameters));
    } else {
      xhr.send(parameters);
    }
  } else if (requestType === 'get') {
    xhr.send();
  }

  return new Promise((resolve, reject) => {
    xhr.onreadystatechange = () => {
      if (xhr.readyState === XMLHttpRequest.DONE && xhr.status === 200) {
        resolve(xhr.responseText);
      } else if (xhr.readyState === XMLHttpRequest.DONE && (xhr.status == 500 || xhr.status == 503)) {
        reject(xhr.responseText);
      } else if (xhr.readyState === XMLHttpRequest.DONE && xhr.status == 404) {
        reject(xhr.responseText);
      }
    };
    xhr.onerror = () => {
      reject({ error: true, msg: '网络错误', data: null });
    };
  });
}


src/util/htmlNameCode.ts
export const htmlStringObj = {
  160: '&nbsp;',
  34: '&quot;',
  38: '&amp;',
  39: '&apos;',
  60: '&lt;',
  62: '&gt;',
  161: '&iexcl;',
  162: '&cent;',
  163: '&pound;',
  164: '&curren;',
  165: '&yen;',
  166: '&brvbar;',
  167: '&sect;',
  168: '&uml;',
  169: '&copy;',
  170: '&ordf;',
  171: '&laquo;',
  172: '&not;',
  173: '&shy;',
  174: '&reg;',
  175: '&macr;',
  176: '&deg;',
  177: '&plusmn;',
  178: '&sup2;',
  179: '&sup3;',
  180: '&acute;',
  181: '&micro;',
  182: '&para;',
  183: '&middot;',
  184: '&cedil;',
  185: '&sup1;',
  186: '&ordm;',
  187: '&raquo;',
  188: '&frac14;',
  189: '&frac12;',
  190: '&frac34;',
  191: '&iquest;',
  192: '&Agrave;',
  193: '&Aacute;',
  194: '&Acirc;',
  195: '&Atilde;',
  196: '&Auml;',
  197: '&Aring;',
  198: '&AElig;',
  199: '&Ccedil;',
  200: '&Egrave;',
  201: '&Eacute;',
  202: '&Ecirc;',
  203: '&Euml;',
  204: '&Igrave;',
  205: '&Iacute;',
  206: '&Icirc;',
  207: '&Iuml;',
  208: '&ETH;',
  209: '&Ntilde;',
  210: '&Ograve;',
  211: '&Oacute;',
  212: '&Ocirc;',
  213: '&Otilde;',
  214: '&Ouml;',
  215: '&times;',
  216: '&Oslash;',
  217: '&Ugrave;',
  218: '&Uacute;',
  219: '&Ucirc;',
  220: '&Uuml;',
  221: '&Yacute;',
  222: '&THORN;',
  223: '&szlig;',
  224: '&agrave;',
  225: '&aacute;',
  226: '&;',
  227: '&atilde;',
  228: '&auml;',
  229: '&aring;',
  230: '&aelig;',
  231: '&ccedil;',
  232: '&egrave;',
  233: '&eacute;',
  234: '&ecirc;',
  235: '&euml;',
  236: '&igrave;',
  237: '&iacute;',
  238: '&icirc;',
  239: '&iuml;',
  240: '&eth;',
  241: '&ntilde;',
  242: '&ograve;',
  243: '&oacute;',
  244: '&ocirc;',
  245: '&otilde;',
  246: '&ouml;',
  247: '&divide;',
  248: '&oslash;',
  249: '&ugrave;',
  250: '&uacute;',
  251: '&ucirc;',
  252: '&uuml;',
  253: '&yacute;',
  254: '&thorn;',
  255: '&yuml;',
  8226: '&bull;',
  8734: '&infin;',
  8240: '&permil;',
  8901: '&sdot;',
  8224: '&dagger;',
  8212: '&mdash;',
  8869: '&perp;',
  8741: '&par;',
  8364: '&euro;',
  8482: '&trade;',
  945: '&alpha;',
  946: '&beta;',
  947: '&gamma;',
  948: '&delta;',
  949: '&epsilon;',
  950: '&zeta;',
  951: '&eta;',
  952: '&theta;',
  953: '&iota;',
  954: '&kappa;',
  955: '&lambda;',
  956: '&mu;',
  957: '&nu;',
  958: '&xi;',
  959: '&omicron;',
  960: '&pi;',
  961: '&rho;',
  963: '&sigma;',
  964: '&tau;',
  965: '&upsilon;',
  966: '&phi;',
  967: '&chi;',
  968: '&psi;',
  969: '&omega;',
  913: '&Alpha;',
  914: '&Beta;',
  915: '&Gamma;',
  916: '&Delta;',
  917: '&Epsilon;',
  918: '&Zeta;',
  919: '&Eta;',
  920: '&Theta;',
  921: '&Iota;',
  922: '&Kappa;',
  923: '&Lambda;',
  924: '&Mu;',
  925: '&Nu;',
  926: '&Xi;',
  927: '&Omicron;',
  928: '&Pi;',
  929: '&Rho;',
  931: '&Sigma;',
  932: '&Tau;',
  933: '&Upsilon;',
  934: '&Phi;',
  935: '&Chi;',
  936: '&Psi;',
  937: '&Omega;',
};


src/component/convert/fragments/helper/UnicodeHelper.ts
import { htmlStringObj } from '@/util/htmlNameCode';
import iconv from 'iconv-lite';
// var iconv = require('iconv-lite');

type GBCharset = 'GB2312' | 'GBK' | 'GB18030';

/**
 * 判断输入是否为空
 */
function isEmpty(data: string): boolean {
  return !data && data.length === 0;
}

export function stringToUnicode(data: string): string {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  let unicodeString = '';
  for (let i = 0; i < inputData.length; i++) {
    const code = inputData.codePointAt(i);
    unicodeString += `u+${inputData.codePointAt(i).toString(16)}`;
    if (code > 0xffff) {
      i++;
    }
  }
  return unicodeString;
}

/**
 * 判断字符串数组每个元素是否为合法的unicode编码
 */
function isUnicode(unicodeArray: string[]): boolean {
  return unicodeArray.every((item) => !isNaN(Number(`0x${item}`)) && Number(`0x${item}`) <= 0x10ffff);
}

/**
 * 用编码前缀分割数组，如果第一个元素为空，去除第一个元素
 */
function customSplit(data: string, symbol: string) {
  const returnArr = data.split(symbol);
  if (returnArr[0].length === 0) {
    returnArr.shift();
  }
  return returnArr;
}

export function unicodeToString(data: string): string {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  let commonString = '';
  const inputString = inputData.toLowerCase();
  if (inputString.includes('u+')) {
    const unicodeArr = customSplit(inputString, 'u+');
    if (isUnicode(unicodeArr)) {
      unicodeArr.forEach((item) => {
        commonString += String.fromCodePoint(Number(`0x${item}`));
      });
    } else {
      return inputData;
    }
  } else {
    return inputData;
  }
  return commonString;
}

export function stringToUTF8(data: string, removeFlag: boolean = false): string {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  const encoder = new TextEncoder();
  let utf8String = '';
  for (let i = 0; i < inputData.length; i++) {
    const code = inputData.codePointAt(i);
    const str = String.fromCodePoint(code);
    const uint8Array = encoder.encode(str);
    uint8Array.forEach((item) => {
      utf8String += removeFlag ? item.toString(16) : `\\x${item.toString(16)}`;
    });
    if (code > 0xffff) {
      i++;
    }
  }
  return utf8String;
}

/**
 * 将16进制字符串转为十进制数字，非法字符串算作0
 */
function hexCodeToDecNumber(data: string): number {
  let hexNumber = 0;
  if (data.includes('0x')) {
    hexNumber = Number(data);
  } else {
    hexNumber = Number(`0x${data}`);
  }
  if (isNaN(hexNumber)) {
    return 0;
  } else {
    return Number(hexNumber.toString(10));
  }
}

/**
 * 将16进制字符串转为十进制数字数组
 */
function getHexToDecArray(data: string): number[] {
  const returnArr = [];
  if (data.length < 2) {
    return [hexCodeToDecNumber(`0${data}`)];
  } else if (data.length !== 2) {
    for (let i = 0; i < data.length; i++) {
      const substring = data.substring(i, i + 2);
      if (substring.length === 2) {
        returnArr.push(hexCodeToDecNumber(substring));
        i++;
      } else {
        returnArr.push(hexCodeToDecNumber(`0${substring}`));
      }
    }
  }
  return returnArr;
}

export function uTF8ToString(data: string) {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  let commonString = '';
  const inputValue = inputData.toLowerCase();
  const decoder = new TextDecoder();
  if (inputValue.includes('\\x')) {
    const utf8Arr = customSplit(inputValue, '\\x');
    const decArr = [];
    utf8Arr.forEach((item) => {
      if (item) {
        if (item.length == 2) {
          decArr.push(hexCodeToDecNumber(`0x${item}`));
        } else {
          decArr.push(...getHexToDecArray(item));
        }
      }
    });
    const bufferArr = new Uint8Array(decArr);
    commonString = decoder.decode(bufferArr);
  } else {
    const decArr = getHexToDecArray(inputValue);
    const bufferArr = new Uint8Array(decArr);
    commonString = decoder.decode(bufferArr);
  }
  return commonString;
}

/**
 * 在前面填充0
 */
function addZeroToHead(data: string, lastLength: number): string {
  if (data.length === lastLength) {
    return data;
  } else if (data.length < lastLength) {
    let tempData = data;
    for (let i = 0; i < lastLength - data.length; i++) {
      tempData = `0${tempData}`;
    }
    return tempData;
  }
}

export function stringToUTF16(data: string, removeFlag: boolean = false): string {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  let utf16String = '';
  for (let i = 0; i < inputData.length; i++) {
    const code = inputData.codePointAt(i);
    if (code > 0xffff) {
      const utf16ArrString = JSON.stringify(String.fromCodePoint(code).split(''));
      utf16String += utf16ArrString.substring(2, utf16ArrString.length - 2).replace(/","/g, '');
      i++;
    } else {
      utf16String += `\\u${code.toString(16)}`;
    }
  }
  const lastStringArr = [];
  const tempArr = utf16String.split('\\u');
  tempArr.shift();
  tempArr.forEach((item) => {
    lastStringArr.push(addZeroToHead(item, 4));
  });
  utf16String = removeFlag ? lastStringArr.join('') : `\\u${lastStringArr.join('\\u')}`;
  return utf16String;
}

/**
 * 将16进制字符串转为十进制数字，非法字符串直接返回该字符串
 */
function hexCodeToDecNumberOrString(data: string): number | string {
  let hexNumber = 0;
  if (data.includes('0x')) {
    hexNumber = Number(data);
  } else {
    hexNumber = Number(`0x${data}`);
  }
  if (isNaN(hexNumber)) {
    return data;
  } else {
    return Number(hexNumber.toString(10));
  }
}

/**
 * 处理字符串，合法16进制编码直接转换，不合法的直接拼接在后面
 */
function getCommonString(data: string, hexLength: number) {
  let commonString = '';
  if (data.length < hexLength) {
    return data;
  } else {
    let hexArr = [];
    for (let i = 0; i < data.length; i++) {
      const substring = data.substring(i, i + hexLength);
      if (substring.length === hexLength) {
        if (isNumber(hexCodeToDecNumberOrString(substring))) {
          hexArr.push(hexCodeToDecNumberOrString(substring));
        } else {
          commonString += substring;
        }
        i = i + hexLength - 1;
      } else {
        if (hexArr.length > 0) {
          commonString += String.fromCodePoint(...hexArr);
          hexArr = [];
        }
        commonString += substring;
      }
    }
    if (hexArr.length > 0) {
      commonString += String.fromCodePoint(...hexArr);
    }
  }
  return commonString;
}

/**
 * 判断是否为数字，且在Unicode范围内
 */
function isNumber(data: string | number): boolean {
  return !isNaN(Number(data)) && Number(data) <= 0x10ffff;
}

/**
 * 转换合法16进制字符串，直接拼接非法字符串
 */
function convertValidHexConnectInvalid(data: string, symbol: string, codeLength: number, removeFlag: boolean): string {
  let commonString = '';
  if (data.includes(symbol)) {
    const inputArr = data.split(symbol);
    let hexArr = [];
    inputArr.forEach((item) => {
      if (item) {
        if (item.length === codeLength) {
          if (isNumber(`0x${item}`)) {
            hexArr.push(`0x${item}`);
          } else {
            commonString += item;
          }
        } else {
          if (item.length > codeLength) {
            if (isNumber(`0x${item.substring(0, codeLength)}`)) {
              hexArr.push(`0x${item.substring(0, codeLength)}`);
            }
            if (hexArr.length > 0) {
              commonString += String.fromCodePoint(...hexArr);
              hexArr = [];
            }
            if (!isNumber(`0x${item.substring(0, codeLength)}`)) {
              commonString += item.substring(0, codeLength);
            }
            commonString += item.substring(codeLength);
          } else {
            commonString += item;
          }
        }
      }
    });
    if (hexArr.length > 0) {
      commonString += String.fromCodePoint(...hexArr);
    }
  } else {
    if (!removeFlag) {
      return data;
    } else {
      return getCommonString(data, codeLength);
    }
  }
  return commonString;
}

export function uTF16ToString(data: string, removeFlag: boolean = false) {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  const inputValue = inputData.toLowerCase().trim();
  let commonString = '';
  commonString = convertValidHexConnectInvalid(inputValue, '\\u', 4, removeFlag);
  return commonString;
}

export function stringToUTF32(data: string, removeFlag: boolean = false): string {
  const inputData = data.trim();
  let utf32String = '';
  for (let i = 0; i < inputData.length; i++) {
    const code = inputData.codePointAt(i);
    const codeHex = code.toString(16);
    let codeUtf32 = codeHex;
    codeUtf32 = addZeroToHead(codeHex, 8);
    utf32String += removeFlag ? codeUtf32 : `u+${codeUtf32}`;
    if (code > 0xffff) {
      i++;
    }
  }
  return utf32String;
}

export function utf32ToString(data: string, removeFlag: boolean = false) {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  const inputValue = inputData.toLowerCase().trim();
  let commonString = '';
  commonString = convertValidHexConnectInvalid(inputValue, 'u+', 8, removeFlag);
  return commonString;
}

//代替被摈弃的unescape方法
function unescapeForbtoa(str: string) {
  return str.replace(/%([0-9A-F]{2})/g, (match, p1) => {
    return String.fromCharCode(Number(`0x${p1}`));
  });
}

export function stringToBase64(data: string) {
  const inputData = data.trim();
  const urlEncode = encodeURIComponent(inputData);
  const base64String = window.btoa(unescapeForbtoa(urlEncode));
  return base64String;
}

export function base64ToString(data: string) {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  return decodeURIComponent(escape(window.atob(inputData)));
}

export function stringToURLCode(data: string) {
  const inputData = data.trim();
  const urlString = encodeURIComponent(inputData);
  return urlString;
}

export function urlCodeToString(data: string) {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  return decodeURIComponent(inputData);
}

export function stringToDecimal(data: string) {
  const inputData = data.trim();
  let decString = '';
  for (let i = 0; i < inputData.length; i++) {
    const code = inputData.codePointAt(i);
    decString += `${code} `;
    if (code > 0xffff) {
      i++;
    }
  }
  return decString.trim();
}

export function decimalToString(data: string) {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  const inputValue = inputData.trim();
  const inputArr = inputValue.split(/\s+/);
  const decArr = [];
  inputArr.forEach((item) => {
    if (Number(item) > 1114111) {
      decArr.push(1114111);
    } else {
      decArr.push(Number(item));
    }
  });
  return String.fromCodePoint(...decArr);
}

export function stringToHTMLCode(data: string): string {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  let htmlCodeString = '';
  const decString = stringToDecimal(inputData);
  decString.split(/\s/).forEach((item) => {
    htmlCodeString += `&#${item}; `;
  });
  return htmlCodeString.trim();
}

export function htmlCodeToString(data: string) {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  const decArr = [];
  const reg = /(?<=&#)(\d+)(?=;)/g;
  const arr = inputData.match(reg) ?? [];

  arr.forEach((item) => {
    if (isNumber(item)) {
      decArr.push(item);
    }
  });
  return decimalToString(decArr.join(' '));
}

export function decNumberToHTMLNameCode(data: string): string {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  let htmlNameString = '';
  const decArr = inputData.split(/\s+/) ?? [];
  decArr.forEach((item) => {
    if (htmlStringObj[item]) {
      htmlNameString += htmlStringObj[item];
    }
  });
  return htmlNameString;
}

export function htmlNameCodeToDecString(data: string): string {
  const inputData = data.trim();
  if (isEmpty(inputData)) {
    return '';
  }
  let decString = '';
  const htmlCodeArr = [];
  const reg = /(?<=&)(.+)(?=;)/g;
  const arr = inputData.match(reg) ?? [];

  arr.forEach((item) => {
    htmlCodeArr.push(`&${item};`);
  });
  htmlCodeArr.forEach((item) => {
    for (const key in htmlStringObj) {
      if (htmlStringObj[key] === item) {
        decString += `${key} `;
      }
    }
  });
  return decString.trim();
}

export function stringToGBEncode(data: string, charSet: GBCharset, removeFlag?: boolean): string {
  if (isEmpty(data)) {
    return;
  }
  const strArr = data.split('');
  let returnStr = '';
  strArr.forEach((item) => {
    const uint8Array = iconv.encode(item, charSet);
    let singleCharCode = '';
    uint8Array.forEach((elem) => {
      let hexStr = elem.toString(16);
      if (charSet === 'GB2312' || charSet === 'GBK') {
        if (item !== '?' && elem === 63) {
          hexStr = '?';
        }
      }
      singleCharCode += hexStr;
    });
    returnStr += singleCharCode !== '?' ? `\\x${singleCharCode}` : '?';
  });
  return returnStr;
}

export function GBCodeToString(data: string, charSet: GBCharset): string {
  if (isEmpty(data)) {
    return '';
  }
  let hexStr = data;
  if (data.indexOf('\\x') >= 0) {
    hexStr = data.replace(/\\x/g, '');
  }
  const reg = /[^0-9a-fA-F]/g;
  if (reg.test(hexStr)) {
    return '';
  }
  if (hexStr.length % 2 !== 0) {
    return '';
  }
  const uint8Arr = [];
  for (let i = 0; i < hexStr.length; i += 2) {
    uint8Arr.push(Number(Number(`0x${hexStr.substring(i, i + 2)}`).toString(10)));
  }
  const uint8Array = new Uint8Array(uint8Arr);
  console.log('u8a', uint8Array);
  const decoder = new TextDecoder(charSet.toLowerCase());
  try {
    return decoder.decode(uint8Array);
  } catch (error) {
    return '';
  }
}


webpack.config.js
const path = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');

const rootDir = process.cwd();

module.exports = {
  entry: {
    ok: './src/index.tsx',
  },
  devServer: {
    static: {
      directory: path.join(__dirname, 'static'),
    },
    compress: true,
    port: 8080,
    proxy: {
      '/b2b': {
        target: 'http://109.14.20.45:6636/ux',
        changeOrigin: true,
      },
      '/lxz': {
        target: 'http://109.14.6.243:7777',
        pathRewrite: { '^/lxz': '' },
        changeOrigin: true,
      },
    },
  },
  output: {
    filename: '[name]-[chunkhash].js',
    path: path.resolve(__dirname, 'dist'),
    assetModuleFilename: 'images/[name][ext]',
    publicPath: '/cusys/react/',
  },
  plugins: [
    new HtmlWebpackPlugin({
      title: 'online tools',
      template: 'src/index.html',
      filename: 'index.html',
      chunks: ['ok'],
    }),
    new MiniCssExtractPlugin({
      filename: '[name]-[chunkhash].css',
    }),
  ],
  resolve: {
    // Add `.ts` and `.tsx` as a resolvable extension.
    extensions: ['.ts', '.tsx', '.js'],
    alias: {
      '@': [path.join(rootDir, 'src')],
    },
  },
  module: {
    rules: [
      {
        test: /\.css$/i,
        //use: ["style-loader", "css-loader"],
        use: [
          {
            loader: MiniCssExtractPlugin.loader,
            options: {
              publicPath: '/',
            },
          },
          'css-loader',
        ],
      },
      {
        test: /\.jsx?$/i,
        use: ['babel-loader'],
      },
      {
        test: /\.tsx?$/,
        use: [
          {
            loader: 'ts-loader',
            options: {
              transpileOnly: true,
            },
          },
        ],
      },
      {
        test: /\.png/,
        type: 'asset',
        parser: {
          dataUrlCondition: {
            maxSize: 30 * 1024, // 4kb
          },
        },
      },
    ],
  },
};
